### DISCUSSION TITLE
AST question: exclude subtree

### DISCUSSION THREAD

**Discussion by michael-brade** [taocpp/PEGTL#92] [Q&A] [answered]
I would like to exclude a node including its subtree from the AST (using `parse_tree.hpp`). The node has child types that are used and needed elsewhere in the AST (e.g. an `assignmentExpression`).

I can write a transform like this:
```
template<> struct store_simple<src::withExpression>     : std::true_type {
  static void transform(std::unique_ptr<pegtl::parse_tree::node>& n) {
    n->children.clear();
  }
};
```
which removes the children, but it keeps the empty `withExpression` in the tree. How do I get rid of it? The following solutions don't seem elegant:

* I could rewrite the grammar so that none of the rules used to parse `withExpression` are needed in the AST (a lot of redundancy)
* I could traverse the final tree and delete the nodes
* I could modify all possible parents of `withExpression` to delete the node

But the best way, I think, would be to somehow access the state and drop (pop) the node before it's added to the tree. But I don't see how to do that with the new API.

Thoughts?

On another note: I do like the new transform API, except for the removal of the bool return value of `parse_tree::parse()` because it's not consistent with the normal `parse()`, and because I want to know if the parsing succeeded and only then use the AST. So now I have to infer from the returned tree state object if the parsing succeeded. Therefore I changed that part locally for my project again. Maybe as an alternative you could put the bool into the state, either directly or as an accessor that checks if there are nodes in the state.

**Comment by d-frey** [comment]
I've committed two changes:

* The `transform`-function could call `n.reset();` on its argument, this will remove the whole node (actually it will prevent it from being added to the parent).
* `parse_tree::parse()` now returns a `std::pair<bool, state>`, where the first value is the result from the actual `parse()` call and the second value is the `state` as before, which might be inconsistent if the first value is `false`.

I hope this solves your problems, if so, please close this issue.

**Comment by d-frey** [comment, marked as answer]
FYI: I just refactored the parse tree support and added some initial documentation. Let me know if you have any questions, remarks, etc.

**Comment by michael-brade** [comment]
This is great, I like it! Yes, already your very first commit would have solved the issue, but `reset()` looks more elegant. I will try it on Monday, and I'm quite sure I won't have to reopen this issue :) Thanks!
